### PR TITLE
feat(openai): add GPT Image 2.5 Sunburst and Flare model information

### PR DESCRIPTION
**Pull request by Laisky**
## Summary

Add the official GPT Image 2.5 catalog entries and repair the five original reviewer findings using **behavioral reproduction before production fixes**. Existing model IDs and tariff tables are retained.

### Added model IDs

- `gpt-image-2.5-sunburst`
- `gpt-image-2.5-sunburst-2026-09-08`
- `gpt-image-2.5-flare`
- `gpt-image-2.5-flare-2026-09-08`

Both variants have text/image input and image output, independent alias/snapshot configurations, `size=auto`, `quality=auto`, a 32,000-Unicode-code-point prompt limit, and `n=1..10`. No undocumented context/output-token limit is invented.

Standard USD per million tokens: text input **5**, cached text input **1.25**, image input **8**, cached image input **2**, image output **30**. Equal token rates do not justify copying GPT Image 2 per-image estimates.

## Important: guarded enablement

For each enabled GPT Image 2.5 ID, configure the existing channel `model_configs` field `image.price_per_image_usd` as an **operator-defined advance reservation and incomplete-usage fallback tariff**. It is not an official provider price or an additional render fee. Without a positive, finite, representable reservation, the relay returns `503 image_billing_not_configured` before any paid upstream request.

The reserve scales with count, configured tier, and group multiplier. Actual billable token cost replaces the reserve; unused quota is refunded by a signed adjustment. Missing/ambiguous usage retains the configured fallback and logs its source. A reserve is an advance estimate, **not a hard provider-cost ceiling**; actual usage can exceed it and require an additional debit under existing quota policy.

See [the operational billing note](docs/usage/openai-image25-billing.md) before enabling traffic. A price-only reservation override preserves the model's Image API defaults.

## Reviewer disposition and implementation

| Finding | Verdict | Remediation |
| --- | --- | --- |
| No meaningful quota reservation before dispatch | Confirmed, necessary | Explicit reserve/fallback policy, real pre-debit before HTTP, signed settlement/refunds |
| Missing usage produces a near-free image | Confirmed, necessary | Configured fallback, distinguish fallback from measured token billing |
| Images decoder drops cached tokens | Confirmed, necessary | Preserve aggregate and modality-specific cache details through production JSON conversion |
| Fixed token table ignores channel prices | Confirmed, necessary | Retain effective request-start channel input/cache/image/output pricing through final settlement |
| Character limit enforced as UTF-8 bytes | Confirmed, necessary | Count Unicode code points for this family; test ASCII, CJK, and emoji boundaries |

Additional coverage catches asymmetric cache allocation, totals-only/partial input, duplicate reserve-plus-token fees, and physical-balance refunds. Explicit cache buckets are used instead of proportional guessing for the new family. Incomplete mixed-cache data is treated as fallback, not invented measured usage. Unclassified input totals use the documented text-rate compatibility fallback rather than being silently dropped.

Channel `ratio`, `completion_ratio`, `cached_input_ratio`, and `image.prompt_ratio` remain effective, with existing request-start time windows and usage tiers. The existing schema has no independently configurable cached-image rate: the image multiplier applies to both input buckets.

Two subsequent suggestions to weaken intentionally RED tests were **not applied**: changing the reserve case to bypass the real per-image flag, and deleting guarded-admission/fallback assertions would hide the reproduced defects. Explanations and evidence are posted in the review threads. The new workflow disables persisted checkout credentials; no existing CI gate was weakened.

## Behavioral evidence: RED → GREEN

1. First test-only commits reproduced reservation, fallback, JSON cache handling, and settlement defects through the real controller, a local HTTP server, and SQLite: [RED run](https://github.com/Laisky/one-api/actions/runs/34369588597), head `40587c2`.
2. Production fix `12e3b30` retained those assertions: [GREEN race-enabled run](https://github.com/Laisky/one-api/actions/runs/34371296464).
3. Test-only commit `ffba145` reproduced the remaining channel-pricing and Unicode defects: [RED run](https://github.com/Laisky/one-api/actions/runs/34371849008). Channel override expected 26,000 quota but got 18,500; the valid 32,000-code-point CJK boundary was rejected.
4. Production fix `e0ccdc9` retained all regression cases: **[GREEN race-enabled run](https://github.com/Laisky/one-api/actions/runs/34373063493)**.

The final focused gate executes real repository packages with Go 1.27.1:

```sh
go test -race -count=1 -timeout=5m ./relay/adaptor/openai ./relay/controller -run 'Test(GPTImage25|Image)'
```

The HTTP/SQLite regressions verify zero upstream calls on rejected requests, physical user/token quota before dispatch, final balances, signed refunds, and persisted consumption logs/request costs. They do not merely call pricing helpers with hand-normalized usage.

## Validation scope

- Final focused behavioral/race gate: **passed** on `e0ccdc9`.
- Modern, Berry, and Air frontend tests, dependency vulnerability scan, entity-response guardrail, and Gin-context guardrail: **passed** on the same head at last inspection.
- Full repository Go checks were still running/queued at last inspection; this PR does **not** claim full CI is green.
- Local checks used formatting/syntax validation and an isolated source harness because the local container has Go 1.23.2, not the required Go 1.27.1. The linked GitHub runs provide actual package/HTTP/database evidence.
- No live paid OpenAI request was made. No streaming Images or Responses image-generation-tool implementation is added.

## Official sources

- https://developers.openai.com/api/docs/models
- https://developers.openai.com/api/docs/models/gpt-image-2.5-sunburst
- https://developers.openai.com/api/docs/models/gpt-image-2.5-flare
- https://developers.openai.com/api/docs/guides/image-generation
- https://developers.openai.com/api/reference/resources/images/methods/generate

## Summary by CodeRabbit

- **New Features**
  - Added support for GPT Image 2.5 image-generation models, including Sunburst and Flare variants and dated snapshots.
  - Added token-based usage billing for text and image inputs, cached inputs, and generated images.
  - Added support for quality options including low, medium, high, xhigh, max, and auto.
  - Requests now default to automatic size and quality settings when unspecified.
  - Added support for generating between 1 and 10 images per request.

### CODE REVIEW

File: relay/adaptor/openai/constants_image25.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// Parameters: description identifies the alias or snapshot. Returns: token-only
+// pricing and Image API defaults without undocumented context or output limits.
+func newGPTImage25Config(description string) adaptor.ModelConfig {
+	return adaptor.ModelConfig{
+		Ratio:            5 * billingratio.MilliTokensUsd,
+		CachedInputRatio: 1.25 * billingratio.MilliTokensUsd,
+		CompletionRatio:  30.0 / 5.0, // Image output, not text output.
+		Image: &adaptor.ImagePricingConfig{
+			// Leave PricePerImageUsd and tier multipliers unset: the controller
+			// reconciles actual image-token usage instead of a fixed render fee.
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/adaptor/openai/constants_image25.go:41]: Confirmed with behavioral tests, not just static inspection. The test-only commit 40587c2 reproduced missing admission/pre-debits through the real RelayImageHelper, a local HTTP upstream, and SQLite. Fixed in 12e3b30: GPT Image 2.5 now requires an explicit channel reservation/fallback tariff before dispatch, uses the existing quota pre-consumption path, replaces the reservation with actual token cost, and applies signed refunds. A price-only override preserves Image API defaults. The reserve is operator policy, not an invented OpenAI per-image price or a hard upper bound. Regression run https://github.com/Laisky/one-api/actions/runs/34371296464 passed with -race, including physical user/token balances before HTTP, insufficient-credit rejection, signed refunds, and log/request-cost reconciliation. Guarded enablement is documented in docs/usage/openai-image25-billing.md.

[the remark this answers is not in the document]
Comment (Laisky) [reply relay/adaptor/openai/constants_image25.go:41]: Confirmed and fixed in 12e3b30. Missing, null, empty, input-only, and ambiguous mixed-cache usage are exercised from provider JSON through final settlement. The new family no longer falls back to a one-token estimate: an explicit operator-configured image.price_per_image_usd reserve/fallback is required before dispatch, retained when billable usage is unavailable, and labeled billing_source=configured_fallback. Reliable token usage replaces the reserve rather than adding a render fee. No GPT Image 2 per-image estimates were copied. The real HTTP/SQLite race-enabled regression run passed: https://github.com/Laisky/one-api/actions/runs/34371296464.

File: relay/controller/image.go [hunk 2]
```diff
@@ -501,6 +501,37 @@ var gptImageTokenBucketPrices = map[string]gptImageTokenBucketPricing{
...
 		outputImageUSD:      30.0,
 	},
+	// GPT Image 2.5 token rates verified 2026-09-09 (not per-image estimates).
+	// https://developers.openai.com/api/docs/models/gpt-image-2.5-sunburst
+	// https://developers.openai.com/api/docs/models/gpt-image-2.5-flare
+	"gpt-image-2.5-sunburst": {
+		inputTextUSD:        5.0,
+		cachedInputTextUSD:  1.25,
+		inputImageUSD:       8.0,
+		cachedInputImageUSD: 2.0,
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/controller/image.go]: Confirmed from actual Images API JSON: the pre-fix decoder discarded cached_tokens, so manually normalized tests were insufficient. Fixed in 12e3b30 by carrying cached_tokens and cached_tokens_details.{text_tokens,image_tokens} through ImageUsage.Convert2GeneralUsage. New-family billing uses explicit cache modalities instead of proportional guessing; ambiguous mixed-cache usage retains the configured fallback. Regression cases include cached text, cached images, unequal/asymmetric cache buckets, uncached controls, totals-only input, and partial modality details. The tests failed before the fix and passed in the real package/HTTP/SQLite race-enabled run: https://github.com/Laisky/one-api/actions/runs/34371296464.

File: relay/controller/image25_review_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	}
+}
+
+// TestGPTImage25ReviewReservationIsNotAnExtraFee verifies token settlement with
+// an operator-configured reserve. Parameters: t runs the assertion. Returns: none.
+func TestGPTImage25ReviewReservationIsNotAnExtraFee(t *testing.T) {
+	t.Parallel()
+	var upstream openai.ImageUsage
+	require.NoError(t, json.Unmarshal([]byte(`{"input_tokens":1000,"output_tokens":1000,"input_tokens_details":{"text_tokens":1000,"image_tokens":0}}`), &upstream))
+	got := finalizeImageQuota(50_000, true, "gpt-image-2.5-sunburst", "gpt-image-2.5-sunburst", upstream.Convert2GeneralUsage(), 1)
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/controller/image25_review_test.go:54]: Not applying this suggested test change. This was an intentional RED behavioral regression: a configured reservation currently sets perImageBilling=true in the real controller, but a reservation must not become an additional render fee. Changing the argument to false would bypass precisely the production path under review and hide the defect. Commit 12e3b30 fixes production settlement for this family while preserving the assertion and true argument. The accompanying real HTTP/SQLite test verifies both the signed balance refund and persisted cost, and the race-enabled suite passed: https://github.com/Laisky/one-api/actions/runs/34371296464.

File: relay/controller/image25_relay_review_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+		{"unconfigured_sunburst_snapshot", "gpt-image-2.5-sunburst-2026-09-08", `{}`, 0, 1_000_000, 1, 503, 0, 0, false},
+		{"unconfigured_flare", "gpt-image-2.5-flare", `{}`, 0, 1_000_000, 1, 503, 0, 0, false},
+		{"unconfigured_flare_snapshot", "gpt-image-2.5-flare-2026-09-08", `{}`, 0, 1_000_000, 1, 503, 0, 0, false},
+		{"insufficient_credit", "gpt-image-2.5-sunburst", cachedUsage, 0.1, 3, 1, 403, 0, 0, false},
+		{"reserve_then_refund_unused", "gpt-image-2.5-sunburst", cachedUsage, 0.1, 1_000_000, 2, 200, 100_000, 18_500, false},
+		{"charge_above_reserve", "gpt-image-2.5-flare", `{"input_tokens":1000,"output_tokens":10000}`, 0.1, 1_000_000, 1, 200, 50_000, 152_500, false},
+		{"missing_usage", "gpt-image-2.5-flare", `null`, 0.1, 1_000_000, 1, 200, 50_000, 50_000, true},
+		{"empty_usage", "gpt-image-2.5-sunburst", `{}`, 0.1, 1_000_000, 1, 200, 50_000, 50_000, true},
+		{"ambiguous_cache_split", "gpt-image-2.5-flare", `{"input_tokens":2000,"output_tokens":1000,"input_tokens_details":{"text_tokens":1000,"image_tokens":1000,"cached_tokens":1000}}`, 0.1, 1_000_000, 1, 200, 50_000, 50_000, true},
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/controller/image25_relay_review_test.go:48]: These expectations intentionally describe the required behavior before its implementation, not the old implementation's behavior. The user requested reproduction before fixes. Replacing new-model IDs with an unknown model, accepting additive reserve fees, guessing cache splits, or removing fallback-source assertions would erase the reproduced defects. Production was fixed in 12e3b30 and this same HTTP/SQLite suite then passed with -race: https://github.com/Laisky/one-api/actions/runs/34371296464. The admission policy requires an explicit operator reserve/fallback because token-price metadata alone cannot provide a safe render reservation or a missing-usage charge. This distinction is documented in docs/usage/openai-image25-billing.md. Assertions retained.

File: relay/controller/image.go [hunk 5]
```diff
@@ -501,6 +501,37 @@ var gptImageTokenBucketPrices = map[string]gptImageTokenBucketPricing{
...
+		outputImageUSD:      30.0,
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/controller/image.go]: Confirmed with a new test-only commit ffba145 and the real HTTP/SQLite path. In https://github.com/Laisky/one-api/actions/runs/34371849008 the configured channel should charge 26,000 quota but charged the fixed-table 18,500. Fixed in e0ccdc9: request-start effective channel pricing is retained through final settlement, including input/output ratios, cached-input pricing, image.prompt_ratio, time-window selection and usage tiers. Existing schema limitations are explicit: image.prompt_ratio scales both image-input buckets; there is no independent cached-image price field. The same regression is now green with -race: https://github.com/Laisky/one-api/actions/runs/34373063493. The test verifies persisted request/log costs and physical user/token balances, not only a pricing helper.

File: relay/adaptor/openai/constants_image25.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+			PromptRatio:      8.0 / 5.0,
+			DefaultSize:      "auto",
+			DefaultQuality:   "auto",
+			PromptTokenLimit: 32000, // Image API prompt-length limit; not context tokens.
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/adaptor/openai/constants_image25.go:45]: Confirmed and necessary. Test-only commit ffba145 adds ASCII/CJK/emoji boundary cases across all four IDs: 32,000 code points must fit and 32,001 must fail. The CJK case reproduced the byte-count bug in https://github.com/Laisky/one-api/actions/runs/34371849008. Fixed in e0ccdc9 by using utf8.RuneCountInString for this family while leaving other models' existing limit semantics unchanged. The unchanged boundary assertions now pass in the real race-enabled package suite: https://github.com/Laisky/one-api/actions/runs/34373063493.

File: .github/workflows/image-billing-regression.yml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  cancel-in-progress: true
+
+jobs:
+  image-billing:
+    name: GPT Image billing behavior
+    runs-on: ubuntu-latest
+    timeout-minutes: 15
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v7.0.1
```
[the remark this answers is not in the document]
Comment (Laisky) [reply .github/workflows/image-billing-regression.yml:32]: Accepted as a small, necessary hardening change to the new workflow. The pre-fix job log confirms checkout used persist-credentials:true. Commit e0ccdc9 explicitly sets persist-credentials:false while retaining pull_request and contents:read; the subsequent checkout, dependency setup, and behavioral regression job all passed: https://github.com/Laisky/one-api/actions/runs/34373063493. No existing CI gates or permissions were relaxed.

File: relay/controller/image25_relay_review_test.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+				if model.DB.First(&user, 1).Error != nil || model.DB.First(&token, 1).Error != nil {
+					upstreamDBError.Store(true)
+				}
+				atUpstreamUser.Store(user.Quota)
+				atUpstreamToken.Store(token.RemainQuota)
+				w.Header().Set("Content-Type", "application/json")
+				_, err := fmt.Fprintf(w, `{"created":1,"data":[{"b64_json":"AQ=="}],"usage":%s}`, tc.usage)
+				if err != nil {
+					upstreamDBError.Store(true)
+				}
```
[the remark this answers is not in the document]
Comment (Laisky) [reply relay/controller/image25_relay_review_test.go:81]: The shared flag could be split to improve failure diagnostics, but this is not a production-accounting defect or a false-positive source in the current suite: require.False(upstreamDBError.Load()) already runs before either quota assertion and aborts the subtest on a read/write failure. The successful real HTTP/SQLite run establishes that neither failure occurred in the reproduced/fixed cases. Deferring this diagnostic-only refactor to keep the requested remediation focused on behavioral correctness; no assertions were weakened.
